### PR TITLE
Create a tutorial for the library

### LINKED ISSUE #75: Library documentation
The library needs comprehensive documentation.

- Everything in the public interface (i.e. types, functions, traits) should contain documentation docstrings.
- We need a tutorial/guide introducing new users to the library and its features. It can be just a single .md file. The main page of the crate's documentation needs to contain a link to this tutorial/guide.

### PR DESCRIPTION
**Pull request by procrastinationfighter**
Fixes #75 
This PR adds a file tutorial.md which contains a short tutorial about how to use our library. It also fills missing documentation, adds the project's license and a readme.

### CODE REVIEW

**Comment by procrastinationfighter** [comment]
@piodul

File: README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+## Examples
+The repository also contains two exemplary applications:
+
+- [Printer](scylla-cdc-printer): prints all changes from CDC log for a selected table.
+- [Replicator](scylla-cdc-replicator): replicates a table from one Scylla cluster to another by reading the CDC log.
+
+## Contact
+TODO
+
```
Comment (procrastinationfighter) [inline README.md:31]: What should we put here? Scylla-cdc-java directs to the `#cdc` channel on the slack.

Comment (piodul) [reply README.md:34]: Yes, it is a good idea to direct to the same slack channel here.

File: tutorial.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    async fn new_consumer(&self) -> Box<dyn Consumer> {
+        Box::new(TutorialConsumer)
+    }
+}
+```
+
+### Adding shared state to the consumer
+A struct implementing the `Consumer` trait should also implement `Send` trait 
+and a struct implementing the `ConsumerFactory` trait should implement `Send` and `Sync` traits.
+Luckily, Rust implements these traits by default if all member variables of a struct implement them.
```
Comment (piodul) [inline tutorial.md:60]: For context, as to _why_ those traits are required, we should mention that the consumer logic is run in separate tokio tasks and the runtime may schedule them on separate threads.

File: tutorial.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+If the consumers need to share some state, like a reference to an object, 
+they can be wrapped in an [Arc](<https://doc.rust-lang.org/std/sync/struct.Arc.html>).
```
Comment (piodul) [inline tutorial.md:63]: An example of this would be nice here in order to illustrate the idea better. For example, you can add a shared counter which indicates the number of rows read so far, and use that number when printing to indicate the "ID" of the row.

We should also include a standard warning here that mutable state should be avoided as it requires synchronization (atomics, mutexes, etc.) which in turn reduces the speedup offered by tokio running the consumer logic on multiple cores.

File: tutorial.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+_________
+_________
+Time UUID: cc07e47a-d07c-11ec-ebc7-e1924abf3e16, operation type: RowUpdate
+pk: 1, ck: 2
+v: Some(8), was deleted: false
+vs: None, was deleted: false, deleted elements: []
+_________
+```
+
+<b>Note:</b> It is not guaranteed that the `CDCRow` struct will be alive after the function is terminated,
```
Comment (piodul) [inline tutorial.md:196]: I think it could be phrased better. It's not that "it's not guaranteed that the CDCRow will be alive" - the CDCRow's lifetime _does not allow it to exist_ after the function terminates.

File: tutorial.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        println!("Hello, scylla-cdc!");
+        Ok(())
+    }
+}
+```
+
+We'll cover on how to use the `CDCRow` structure in [Using CDCRow](tutorial.md#Using CDCRow).
+
+The library is going to create many instances of `TutorialConsumer`, so we also need to define
+a `ConsumerFactory` for them:
```
Comment (piodul) [inline tutorial.md:45]: IMO it would be good to mention that the library creates a separate consumer for each cdc stream.

**Review by piodul — CHANGES_REQUESTED** [review]
A topic not really explored in our documentation but which I think is worth explaining is the algorithm the library uses to read the CDC log. So, we should explain things like:

- The fact that the library reads one generation at a time, and reads all streams from that generation concurrently,
- Lifetime of the consumers (when they are created and when they are destroyed),
- How the window sizes, delays, etc. relate to how a single stream is being read.

I'm not sure what would be the best place (or places?) to put it, so I leave the choice up to you.

**Comment by procrastinationfighter** [comment]
Besides applying your suggestions, I also added some module-level docstrings for the modules `stream_reader`, `log_reader` and `consumer`.

File: tutorial.md [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        counter.add(1);
+        println!("Row no.{}", counter.get());
+        Ok(())
+    }
+}
+```
+(an external crate [atomic_counter](https://docs.rs/atomic-counter/latest/atomic_counter/) was used in this example)
+
+<b>Note:</b> in general, keeping a mutable state in the `Consumer` is not recommended, 
+since it requires synchronization (i.e. a mutex),
```
Comment (piodul) [inline tutorial.md:91]: I'm guessing that you added a mutex to the `CountingConsumer` so that this statement makes more sense? Atomics are more costly than non-atomic integers, especially if they are accessed from multiple threads at once, so the advice applies to them, too.

File: tutorial.md [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+vs: None, was deleted: false, deleted elements: []
+_________
+```
+
+<b>Note:</b> CDCRow's lifetime does not allow it to exist after the function terminates.
+If the user wants to save the data for later 
+(e.g. in a consumer that saves all consumed rows in a `vec`)
+they should map the rows to another data structure.
+## Saving progress
+TODO
```
Comment (piodul) [inline tutorial.md:230]: Please collaborate with @macher259 on this - I know that you are busy with the benchmarks right now, so maybe he can help you write this section if he has some free time to spare.

You can wait with this until https://github.com/piodul/scylla-cdc-rust/pull/69 is merged, or at least until you are fairly sure that the interface/semantics won't change further during the review.

File: tutorial.md [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        .end_timestamp(end)
+        .consumer_factory(factory)
+        .build()
+        .await
+        .expect("Creating the log reader failed!");
+
+    handle.await
+}
+```
+[//]: # (TODO: Add link to log reader builder's API documentation)
```
[the remark this answers is not in the document]
Comment (procrastinationfighter) [reply tutorial.md:133]: Well, we haven't published the docs yet, so I'm not sure how to link to this now.

File: tutorial.md [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+- to check if a column was deleted in given operation, use `is_value_deleted`
+- to check which elements were deleted from a collection, use `get_deleted_elements` or `take_deleted_elements`
+
+The data returned is of type [CqlValue](<https://docs.rs/scylla/latest/scylla/frame/response/result/enum.CqlValue.html>)
+from the driver.
+
+For more detailed information about how to interpret these values, 
+check the [CDC documentation](<https://docs.scylladb.com/using-scylla/cdc/cdc-log-table/>).
+
+[//]: # (TODO: Add link to the API documentation)
```
[the remark this answers is not in the document]
Comment (procrastinationfighter) [reply tutorial.md:177]: Same as the comment about log reader's docs - I'm referring here to the crates docs that will be available once we've published the library. I'm not sure how to link to this now.

**Comment by procrastinationfighter** [comment]
@macher259 has created the chapter about saving progress, I guess this PR is ready to be reviewed and merged.

**Comment by procrastinationfighter** [comment]
I tried adding the links to the docs - it's not guaranteed that they'll work, but it seems fine. I also deleted the `installation` section from readme.

File: scylla-cdc/src/consumer.rs [hunk 10]
```diff
@@ -351,8 +351,8 @@ mod tests {
             ],
             1,
         )
-        .await?
-        .0;
+            .await?
+            .0;
```
Comment (piodul) [inline scylla-cdc/src/consumer.rs]: Unnecessary whitespace change

File: tutorial.md [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+_________
+_________
+Time UUID: cc07e47a-d07c-11ec-ebc7-e1924abf3e16, operation type: RowUpdate
+pk: 1, ck: 2
+v: Some(8), was deleted: false
+vs: None, was deleted: false, deleted elements: []
+_________
+```
+
+<b>Note:</b> CDCRow's lifetime does not allow it to exist after the function terminates.
```
Comment (piodul) [inline tutorial.md:222]: Nitpick: in Markdown, you can mark text as bold by putting it `__between double underscores__`.

File: scylla-cdc/src/consumer.rs [hunk 12]
```diff
@@ -58,6 +74,8 @@ const TTL_NAME: &str = "cdc$ttl";
 const IS_DELETED_PREFIX: &str = "cdc$deleted_";
 const ARE_ELEMENTS_DELETED_PREFIX: &str = "cdc$deleted_elements_";
 
+/// An internal structure used to map names of columns in the CDC
+/// to their indices in an internal vector.
```
Comment (piodul) [inline scylla-cdc/src/consumer.rs:78]: In what sense it is internal? It is a part of the public API.

File: tutorial.md [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+There is a default implementation that uses ScyllaDB called `TableBackedCheckpointSaver`. 
+Its source code is located in file [checkpoints.rs](scylla-cdc/src/checkpoints.rs) and can serve as an inspiration how to write new checkpoint savers.
+
+Example of usage for `TableBackedCheckpointSaver`:
+```rust
+let user_checkpoint_saver = Arc::new(
+            TableBackedCheckpointSaver::new_with_default_ttl(session, "ks", "checkpoints")
+                .await
+                .unwrap(),
+        );
```
Comment (piodul) [inline tutorial.md:264]: These lines are indented too much with compared to the first one. The last two snippets have the same problem, too.

File: scylla-cdc/src/log_reader.rs [hunk 14]
```diff
@@ -1,3 +1,20 @@
+//! A module containing the logic behind reading and consuming the data.
+//!
+//! Only one stream generation is being read at a time,
+//! while data from each stream is being read concurrently by different StreamReaders.
+//!
+//! The reading is done by reading all CDC rows added in a time window
+//! (not longer than `window_size`), then waiting for amount of time equal to `sleep_interval`
+//! before reading the next window.
+//! Before a newly added row is being read by the reader,
+//! a set amount of time, represented by `safety_interval`, must pass.
```
Comment (piodul) [inline scylla-cdc/src/log_reader.rs:10]: Please rephrase this section so that it doesn't describe "the reading" but rather what a `StreamReader` is doing. Otherwise it might suggest that stream readers are synchronized in some way and they read a time window at the same time which isn't necessarily true.

File: tutorial.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (piodul) [reply tutorial.md:174]: You already added some guessed links, so what about this line?

File: scylla-cdc/src/consumer.rs (hunk 12 above)
Comment (procrastinationfighter) [reply scylla-cdc/src/consumer.rs:78]: To be honest, it shouldn't be public. But we can't create the replication tests without it being public (we don't use the whole library there and we can't create `CDCRowSchema` if it's not public). I will just remove the adjective.

**Comment by piodul** [comment]
Looks good, but there is one important thing missing. When I open the scylla-cdc's crate docs generated by `cargo doc`, I get a list of modules without any explanation:

image

This is because the crate level documentation is missing (no documentation for the main module of scylla-cdc). Some people might get to the docs.rs page skipping github.com and crates.io altogether and will miss the contents of README.md, so we definitely need to put something here (as many other crates do).

Please add documentation to the main module of scylla-cdc. This documentation doesn't have to be long and should tell the user:

- What the crate is about,
- Give a general idea about how to use the crate (e.g. a short example + links to the tutorial).

Please look at the main page of the [scylla-rust-driver](https://docs.rs/scylla/latest/scylla/) for inspiration.

**Comment by procrastinationfighter** [comment]
Done. I also updated some docs, because they linked to the StreamReader, which is a hidden object now.

**Review by piodul — APPROVED** [review]
